src/Storages/ObjectStorage/DataLakes/Iceberg/IcebergWrites.cpp

```diff
@@ -1152,6 +1154,105 @@ std::vector<std::pair<size_t, Field>> DataFileStatistics::getUpperBounds() const
     return result;
 }
 
+MultipleFileWriter::MultipleFileWriter(
+    UInt64 max_size_,
+    Poco::JSON::Array::Ptr schema,
+    FileNamesGenerator & filename_generator_,
+    ObjectStoragePtr object_storage_,
+    ContextPtr context_,
+    const std::optional<FormatSettings> & format_settings_,
+    StorageObjectStorageConfigurationPtr configuration_,
+    SharedHeader sample_block_)
+    : max_size(max_size_)
+    , stats(schema)
+    , filename_generator(filename_generator_)
+    , object_storage(object_storage_)
+    , context(context_)
+    , format_settings(format_settings_)
+    , configuration(configuration_)
+    , sample_block(sample_block_)
+{
+}
+
+void MultipleFileWriter::consume(const Chunk & chunk)
+{
+    if (!current_file_size || *current_file_size >= max_size)
+    {
+        current_file_size = 0;
+        auto filename = filename_generator.generateDataFileName();
+
+        data_file_names.push_back(filename.path_in_storage);
+        auto buffer = object_storage->writeObject(
+            StoredObject(filename.path_in_storage), WriteMode::Rewrite, std::nullopt, DBMS_DEFAULT_BUFFER_SIZE, context->getWriteSettings());
+
+        buffers.push_back(std::move(buffer));
+        if (format_settings)
+        {
+            format_settings->parquet.write_page_index = true;
+            format_settings->parquet.bloom_filter_push_down = true;
+            format_settings->parquet.filter_push_down = true;
+        }
+        output_formats.push_back(FormatFactory::instance().getOutputFormatParallelIfPossible(
+            configuration->format, *buffers.back(), *sample_block, context, format_settings));
+    }
+    output_formats.back()->write(sample_block->cloneWithColumns(chunk.getColumns()));
+    *current_file_size += chunk.getNumRows();
+    stats.update(chunk);
+}
+
+void MultipleFileWriter::finalize()
+{
+    for (auto & output_format : output_formats)
+    {
+        output_format->flush();
+        output_format->finalize();
+    }
+    for (auto & buffer : buffers)
+    {
+        buffer->finalize();
+    }
+}
+
+void MultipleFileWriter::release()
+{
+    for (auto & output_format : output_formats)
+    {
+        output_format.reset();
+    }
+    for (auto & buffer : buffers)
+    {
+        buffer.reset();
+    }
+}
+
+void MultipleFileWriter::cancel()
+{
+    for (auto & output_format : output_formats)
+    {
+        output_format->cancel();
+    }
+    for (auto & buffer : buffers)
+    {
+        buffer->cancel();
+    }
+}
+
+void MultipleFileWriter::clearAllDataFiles() const
+{
+    for (const auto & data_filename : data_file_names)
+        object_storage->removeObjectIfExists(StoredObject(data_filename));
+}
+
+UInt64 MultipleFileWriter::getResultBytes() const
+{
+    UInt64 result = 0;
+    for (auto & buffer : buffers)
+    {
+        result += buffer->count();
```

Comment:
Could be easier to just calculate total inserted bytes in `consume` (as chunks arrive) and here just return the already calculated value